src/vfs2.c

```diff
@@ -472,39 +471,37 @@ static int wal_swap(struct entry *e, const struct wal_hdr *wal_hdr)
 	sqlite3_file *phys_incoming = e->wal_prev;
 	char *name_incoming = e->wal_prev_fixed_name;
 
-	/* Write the new header of the incoming WAL. */
-	rv = phys_incoming->pMethods->xWrite(phys_incoming, wal_hdr,
-					     sizeof(struct wal_hdr), 0);
-	if (rv != SQLITE_OK) {
-		return rv;
-	}
-
-	/* In-memory WAL swap. */
 	e->wal_cur = phys_incoming;
 	e->wal_cur_fixed_name = name_incoming;
 	e->wal_prev = phys_outgoing;
 	e->wal_prev_fixed_name = name_outgoing;
 	e->wal_cursor = 0;
 	e->wal_prev_hdr = e->wal_cur_hdr;
-	e->wal_cur_hdr = *wal_hdr;
+	e->wal_cur_hdr = *hdr;
 
-	/* Move the moving name. */
+	/* Best-effort: flip the moving name.
+	 *
+	 * If these syscalls fail, we can end up with no moving name, or a
+	 * moving name that points to the wrong WAL. We don't use the moving
+	 * name as the source of truth, so this can't lead to dqlite operating
+	 * incorrectly. At worst, it's inconvenient for users who want to
+	 * inspect their database with SQLite (readonly! when dqlite is not
+	 * running!). */
 	rv = unlink(e->wal_moving_name);
-	if (rv != 0 && errno != ENOENT) {
-		return SQLITE_IOERR;
-	}
+	(void)rv;
 	rv = link(name_incoming, e->wal_moving_name);
-	if (rv != 0) {
-		return SQLITE_IOERR;
-	}
+	(void)rv;
 
-	/* TODO do we need an fsync here? */
+	/* Best-effort: invalidate the header of the outgoing physical WAL, so
+	* that it can't be mistakenly applied to the database.
 
-	/* Best-effort: invalidate the outgoing physical WAL so that nobody gets
-	 * confused. */
+	* This provides some protection against users manipulating the
+	* database with SQLite, or a bug in dqlite. But we don't rely on it
+	* for correctness. */
 	(void)phys_outgoing->pMethods->xWrite(phys_outgoing, &invalid_magic,
 					      sizeof(invalid_magic), 0);
-	return SQLITE_OK;
+
+	/* TODO do we need an fsync here? */
```

Comment:
This came up in a discussion with @just-now. I agree the fsync is not needed for the sake of the operations in this function, the idea was that it might be needed to avoid writes to the incoming WAL from being reordered before writes to the outgoing WAL. Suppose we had this order of events on some DB thread:

1. write `tx-A` to WAL1
2. checkpoint WAL1
3. WAL swap
4. write `tx-B` to WAL2

If we crashed and the data from (4) made it to WAL2 but the data from (1) didn't make it to WAL1, we'd have a problem.

However! SQLite does its own fsyncs, and this includes syncing the WAL before doing a checkpoint. That means that our (2) has `fsync(WAL1)` built in, so we don't need to do the same operation ourselves as part of (3). And since a write to the outgoing WAL and a write to the incoming WAL are always separated by a checkpoint, this generalizes. So I think we don't need an fsync and this comment can be removed (but Anatoliy will sanity-check me).